aptos-move/mvhashmap/src/unsync_map.rs

```diff
@@ -0,0 +1,145 @@
+}
+
+impl<K: ModulePath + Hash + Clone + Eq + Ord, V: TransactionWrite + Clone, X: Executable> Default
+    for UnsyncMap<K, V, X>
+{
+    fn default() -> Self {
+        Self {
+            data_map: RefCell::new(BTreeMap::new()),
+            executable_cache: RefCell::new(HashMap::new()),
+            executable_bytes: RefCell::new(0),
+        }
+    }
+}
+
+impl<K: ModulePath + Hash + Clone + Eq + Ord, V: TransactionWrite + Clone, X: Executable>
+    UnsyncMap<K, V, X>
+{
+    pub fn fetch_data(&self, key: &K) -> Option<V> {
+        self.data_map
+            .borrow()
+            .get(key)
+            .and_then(|entry| entry.data.clone())
+    }
+
+    pub fn fetch_code(&self, key: &K) -> Option<MVCodeOutput<V, X>> {
+        use MVCodeOutput::*;
+        debug_assert!(key.module_path().is_some());
+
+        self.data_map
+            .borrow_mut()
+            .get_mut(key)
+            .and_then(|entry| match &entry.data {
+                Some(module) => {
+                    let hash = entry.hash.get_or_insert(module_hash(module));
+
+                    Some(self.executable_cache.borrow().get(hash).map_or_else(
+                        || Module((module.clone(), *hash)),
+                        |x| Executable((x.clone(), ExecutableDescriptor::Published(*hash))),
+                    ))
+                },
+                None => entry
+                    .maybe_base_executable
+                    .as_ref()
+                    .map(|x| Executable((x.clone(), ExecutableDescriptor::Storage))),
+            })
+    }
+
+    pub fn insert(&self, key: K, value: V) {
+        self.data_map
+            .borrow_mut()
+            .insert(key, Entry::from_data(value));
+    }
+
+    /// We return false if the executable was already stored, as this isn't supposed to happen
+    /// during sequential execution (and the caller may choose to e.g. log a message).
+    pub fn store_executable(
+        &self,
+        key: &K,
+        descriptor: ExecutableDescriptor,
+        executable: X,
+    ) -> bool {
+        let size = executable.size_bytes();
+        let ret = match descriptor {
+            ExecutableDescriptor::Published(hash) => self
+                .executable_cache
+                .borrow_mut()
+                .insert(hash, executable)
+                .is_none(),
+            ExecutableDescriptor::Storage => match self.data_map.borrow_mut().entry(key.clone()) {
+                Occupied(mut entry) => entry
+                    .get_mut()
+                    .maybe_base_executable
+                    .replace(executable)
+                    .is_none(),
+                Vacant(entry) => {
+                    entry.insert(Entry::from_base_executable(executable));
+                    true
+                },
+            },
+        };
+
+        if ret {
+            *self.executable_bytes.borrow_mut() += size;
+        }
+
+        ret
+    }
+
+    pub fn executable_size(&self) -> usize {
```

Comment:
Why do we need this for sequential?

Reply:
for tracking memory consumption, we can log it at the end of the block.